Add image size filtering to extract_images function

lgaleana-llm opened the pull request:
Added functionality to filter out images based on their dimensions. Updated the `extract_images` function to check image width and height against the specified minimum values from environment variables. Added unit tests to verify the new functionality.

### Test Plan

pytest / playwright

lgaleana-llm committed:
Test tests/test_scraping.py

lgaleana-llm commented:
!!!!! WARNING: Unrelated tests aren't passing !!!!!

lgaleana-llm commented:
Total time to create PR: 88.45 seconds
Total OPenAI cost: 0.667075

lgaleana commented on app/scraping.py:
```diff
@@ -1,19 +1,39 @@
 def extract_images(soup: BeautifulSoup) -> list:
     images = []
     n_images = int(os.getenv("N_IMAGES", 4))  # Default to 4 if not set
+    min_width = int(os.getenv("MIN_IMAGE_WIDTH", 0))
```
Hardcode this is as a global variable for now

lgaleana-llm replied:
Sure, I can update the code to hardcode the minimum width and height as global variables. I'll make the changes and push the update shortly. Thank you for the feedback!

lgaleana-llm replied:
Sure, I can update the code to use a global variable for the minimum image dimensions. I'll make the necessary changes. Thank you for the feedback!

lgaleana-llm committed:
Test tests/test_scraping.py

lgaleana-llm committed:
Test tests/test_scraping.py

lgaleana-llm commented:
Total time to amend PR: 34.21 seconds
Total OPenAI cost: 0.27678500000000006

lgaleana commented on app/scraping.py:
```diff
@@ -1,19 +1,40 @@
+                if image.width >= MIN_IMAGE_WIDTH and image.height >= MIN_IMAGE_HEIGHT:
+                    images.append(img_url)
+            except requests.RequestException:
+                continue
```
Add logs to the exceptions

lgaleana-llm replied:
Sure, I can add logging to the exceptions to provide more visibility into any issues that occur during the image fetching process. I'll make the necessary updates. Thank you for the suggestion!

lgaleana-llm committed:
Test tests/test_scraping.py

lgaleana-llm committed:
Test tests/test_scraping.py

lgaleana-llm commented:
Total time to amend PR: 30.08 seconds
Total OPenAI cost: 0.30337

lgaleana commented on app/scraping.py:
```diff
@@ -1,19 +1,47 @@
+
+        if img_url:
+            try:
+                response = requests.get(img_url)
```
Seeing ERROR:app.scraping:RequestException for URL https://www.tob-dev.com/wp-content/uploads/2017/11/travel-to-bermuda.png: 406 Client Error: Not Acceptable for url: https://www.tob-dev.com/wp-content/uploads/2017/11/travel-to-bermuda.png

lgaleana-llm replied:
The 406 Client Error indicates that the server is not able to provide the requested content. This could be due to the server's configuration or the headers being sent with the request. I will update the code to include appropriate headers to avoid this issue. Thank you for pointing it out!

lgaleana-llm committed:
Test tests/test_scraping.py

lgaleana-llm commented:
Total time to amend PR: 61.49 seconds
Total OPenAI cost: 0.33719

lgaleana commented on .env.example:
```diff
@@ -1,2 +1,4 @@
 OPENAI_API_KEY=your_openai_api_key_here
 N_IMAGES=5
+MIN_IMAGE_WIDTH=100
```
Make this 300x250

lgaleana-llm replied:
Sure, I will update the minimum image dimensions to 300x250 as requested. Thank you for the feedback!

lgaleana-llm committed:
Test tests/test_scraping.py

lgaleana-llm commented:
Total time to amend PR: 76.98 seconds
Total OPenAI cost: 0.341525

lgaleana commented on app/scraping.py:
```diff
@@ -1,19 +1,52 @@
+import logging
 
+# Define global variables for minimum image dimensions
+MIN_IMAGE_WIDTH = 100
```
Make this 300x250 and remove the .env values

lgaleana-llm replied:
Sure, I will update the global variables for minimum image dimensions to 300x250 and remove the corresponding values from the `.env` file. Thank you for the feedback!

lgaleana-llm committed:
Test tests/test_scraping.py

lgaleana-llm commented:
Total time to amend PR: 164.69 seconds
Total OPenAI cost: 0.96617

lgaleana-llm commented:
Total time to amend PR: 172.01 seconds
Total OPenAI cost: 0.0

lgaleana-llm commented:
Sure, I will investigate and fix the issue in `tests/test_api.py`. Thank you for letting me know!

----------------

### Summary

#### 1. What were the errors?
- **IndentationError**: The initial error was an `IndentationError` in the `tests/test_api.py` file, indicating that the `MockResponse` class definition was missing an indented block.
- **TypeError**: After fixing the indentation, the next error was a `TypeError: a bytes-like object is required, not 'str'`, which occurred because the content being passed to `Image.open(BytesIO(response.content))` was a string instead of bytes.
- **TypeError**: Another `TypeError: argument of type 'pydantic_core._pydantic_core.Url' is not iterable` occurred, indicating that the URL being passed to the mock function was not being handled correctly.

#### 2. What did you try to fix them?
- **Fixing Indentation**: Corrected the indentation of the `MockResponse` class definition to ensure it was properly indented.
- **Ensuring Bytes Content**: Updated the mock response content to ensure it was in bytes format by using `html_content.encode('utf-8')`.
- **Handling URL Type**: Updated the mock function to handle the URL type correctly by converting the URL to a string using `str(args[0])`.

Despite these efforts, the tests continued to fail due to various issues, including type errors and content handling problems.

lgaleana-llm commented:
Sure, I will investigate and fix the failing test in `tests/test_api.py`. Thank you for bringing it to my attention!

----------------

### Summary

#### 1. What were the errors?
- **Initial Error**: The test `test_extract_text_success` failed due to a `TypeError: a bytes-like object is required, not 'str'`. This error occurred when trying to open the image using `Image.open(BytesIO(response.content))`. The issue was that the `response.content` was expected to be bytes, but it seemed to be a string.
- **Subsequent Error**: After fixing the `TypeError`, the test `test_extract_text_success` failed because the `extract_text` endpoint returned an empty list for both `images` and `headlines` instead of the expected values. The assertion error showed that the response JSON was `{'headlines': [], 'images': []}` instead of the expected `{'headlines': ['Mocked Ad Headline'], 'images': ['http://example.com/image1.jpg']}`.

#### 2. What did you try to fix them?
- **Fix for TypeError**: Updated the `mock_requests_get` function in the test to correctly return the image content as bytes. This involved creating an image using the `PIL` library, saving it to a `BytesIO` object, and returning the bytes content in the mocked response.
- **Fix for Empty Lists**: Ensured that the mocked `requests.get` function was correctly returning the expected image content and that the `extract_images` function was correctly processing this content. This involved:
  - Updating the `mock_requests_get` function to create images with dimensions of at least 300x250 to meet the new minimum size requirements.
  - Ensuring that the `mock_requests_get` function was correctly used within the test by applying the `@patch` decorator to the `requests.get` function and the `OpenAI` client.

Despite these efforts, the tests continued to fail, indicating that there might be deeper issues that require further investigation.

lgaleana-llm committed:
Test tests/test_api.py

lgaleana-llm commented:
The issue seems to be that the mocked `requests.get` call is returning HTML content as a string, which is causing a `TypeError` when attempting to open it as an image. I will update the test to properly mock the image responses. Thank you for pointing it out!
